SDL/Kernels.h

```diff
@@ -479,26 +479,26 @@ namespace SDL
                 float eta_pix1 = segmentsInGPU.eta[ix];
                 float phi_pix1 = segmentsInGPU.phi[ix];
 
-                for(int jx = globalThreadIdx[2]; jx < nPixelSegments; jx += gridThreadExtent[2])
+                for(int jx = ix + 1 + globalThreadIdx[2]; jx < nPixelSegments; jx += gridThreadExtent[2])
                 {
                     float eta_pix2 = segmentsInGPU.eta[jx];
                     float phi_pix2 = segmentsInGPU.phi[jx];
 
-                    if (ix == jx)
-                        continue;
-
                     if (alpaka::math::abs(acc, eta_pix2 - eta_pix1) > 0.1f)
                         continue;
 
                     if (secondpass && (!segmentsInGPU.isQuad[jx] || segmentsInGPU.isDup[jx]))
                         continue;
 
                     char quad_diff = segmentsInGPU.isQuad[ix] - segmentsInGPU.isQuad[jx];
-                    float ptErr_diff = segmentsInGPU.ptIn[ix] - segmentsInGPU.ptIn[jx];
                     float score_diff = segmentsInGPU.score[ix] - segmentsInGPU.score[jx];
-                    // Always keep quads over trips. If they are the same, we want the object with the lower pt Error
-                    if((quad_diff > 0) || (score_diff < 0 && quad_diff == 0))
```

Comment:
Hi Ariostas, I am not familiar with the thread competing issue, but I'm curious to learn about this. In the original code, you said that when different thread checking jx in parallel, some thread will find a duplicate, and it "continue" the second loop. But the other thread will still do it's work and check the jx,  right? Can you explain that why it's removing extra things?

Reply:
I think what happens is something like this. Let's say that we have segments with indices 0,1,2, such that this happens:
- when ix=1, jx=0 then 1 is marked as duplicate
- when ix=2, jx=1 then 2 is marked as duplicate
- with all other combinations nothing is marked as duplicate

If we do this process serially, then we start by checking ix=0, and nothing is marked as duplicate, Then we go to ix=1, jx=0 and 1 is marked as duplicate. Then we get to the iteration with ix=2, jx=1, but since 1 was already marked as duplicate then the loop continues. So in the end the segments with indices 0 and 2 remain.

If we do this in parallel, the following can happen. The loop iteration with ix=2, jx=1 can start before the one with ix=1, jx=2 (or at the same time), and since 1 hasn't been marked as duplicate yet then both iterations proceed. These two threads will mark 1 and 2 as duplicate, so in the end only the segment with index 0 will remain.

So the parallel computation can remove extra things, but it seems like the effect is fairly minor, so it's probably acceptable.